### DISCUSSION TITLE
Linking error

### DISCUSSION THREAD

**Discussion by moronmonkey** [DMTF/libspdm#2609] [Q&A]
I recently started getting this error 
test.cpp just contains a function to get revision.
```
g++-9 test.cpp `pkg-config --cflags libspdm` -lspdm -lspdm_crypto  -lspdm_platform -lmbedcrypto
/usr/bin/ld: /tmp/cclCWsB7.ltrans0.ltrans.o: in function `libspdm_generate_challenge_auth_signature':
/home/kargur01/new_DMTF_Repo/libspdm/library/spdm_common_lib/libspdm_com_crypto_service.c:793: undefined reference to `libspdm_requester_data_sign'
/usr/bin/ld: /home/kargur01/new_DMTF_Repo/libspdm/library/spdm_common_lib/libspdm_com_crypto_service.c:813: undefined reference to `libspdm_responder_data_sign'
/usr/bin/ld: /tmp/cclCWsB7.ltrans0.ltrans.o: in function `libspdm_generate_session_handshake_key':
/home/kargur01/new_DMTF_Repo/libspdm/library/spdm_secured_message_lib/libspdm_secmes_session.c:226: undefined reference to `libspdm_psk_handshake_secret_hkdf_expand'
/usr/bin/ld: /home/kargur01/new_DMTF_Repo/libspdm/library/spdm_secured_message_lib/libspdm_secmes_session.c:269: undefined reference to `libspdm_psk_handshake_secret_hkdf_expand'
/usr/bin/ld: /tmp/cclCWsB7.ltrans0.ltrans.o: in function `libspdm_generate_session_data_key':
/home/kargur01/new_DMTF_Repo/libspdm/library/spdm_secured_message_lib/libspdm_secmes_session.c:422: undefined reference to `libspdm_psk_master_secret_hkdf_expand'
/usr/bin/ld: /home/kargur01/new_DMTF_Repo/libspdm/library/spdm_secured_message_lib/libspdm_secmes_session.c:465: undefined reference to `libspdm_psk_master_secret_hkdf_expand'
/usr/bin/ld: /home/kargur01/new_DMTF_Repo/libspdm/library/spdm_secured_message_lib/libspdm_secmes_session.c:509: undefined reference to `libspdm_psk_master_secret_hkdf_expand'
/usr/bin/ld: /tmp/cclCWsB7.ltrans0.ltrans.o: in function `libspdm_get_response_set_certificate':
/home/kargur01/new_DMTF_Repo/libspdm/library/spdm_responder_lib/libspdm_rsp_set_certificate.c:139: undefined reference to `libspdm_is_in_trusted_environment'
/usr/bin/ld: /home/kargur01/new_DMTF_Repo/libspdm/library/spdm_responder_lib/libspdm_rsp_set_certificate.c:162: undefined reference to `libspdm_write_certificate_to_nvm'
/usr/bin/ld: /home/kargur01/new_DMTF_Repo/libspdm/library/spdm_responder_lib/libspdm_rsp_set_certificate.c:219: undefined reference to `libspdm_write_certificate_to_nvm'
```

**Comment by steven-bellock** [comment]
>I recently started getting this error

You updated libspdm or have you been enabling new functionality? Those functions are all behind `#define`s. For example https://github.com/DMTF/libspdm/blob/9bc69639f112be9d5bd8a4aa93cd0fe07d70361a/include/hal/library/responder/setcertlib.h#L15-L28

Ie if you're making a Responder that supports `SET_CERTIFICATE` then it needs to implement `libspdm_is_in_trusted_environment()`.

**Reply by moronmonkey** [reply]
I was facing this issue when I made a Debug build , after changing it to release build its working for 
`libspdm_get_response_certificate`
but same linking error for
`libspdm_get_response_key_exchange`

Is there a need to link` device_secret_lib_sample` library  ??

**Comment by moronmonkey** [comment]
I am not actually changing or implementing any new responder functions , I want to use the default responder functions , there is no device available yet. Which all libraries I need to link ? Linking to libspdm.a , libspdm_crypto.a and libspdm_platform.a gives undefined error for
`libspdm_responder_data_sign`

**Reply by steven-bellock** [reply]
Simplest thing to do is just `cmake -DARCH=x64 -DTOOLCHAIN=GCC -DTARGET=Debug -DCRYPTO=mbedtls ..` and that will include the stuff in https://github.com/DMTF/libspdm/tree/main/os_stub/spdm_device_secret_lib_sample. When/if you are ready to make a real device you will need to provide implementations for those functions.
